### PR TITLE
[FIX] [8.0-i196-BR001028-T22322] Check BR status creating project from sub-BR 

### PR DESCRIPTION
**Pull request by sudhir-serpentcs**
@elicoidal @seb-elico @victormartinelicocorp 
#196 
Fixed  BR status when creating Project from sub-BR

### CODE REVIEW

**Comment by elicoidal** [comment]
@sudhir-serpentcs It is not working as I expected because my issue was not clear enough.

> Expected: a check should verify that the sub-projects are at stakeholder approval stage. if the sub-BR is not at stake holder approval with not linked project, the BR should be ignored in the project creation process.

I meant the following: if the sub-BR is not at stake holder approval with not linked project, the BR should be ignored in the project creation process but any other project should be created (including the parent project).
Besides the button Generate Project should be displayed while parent project and all children has a linked project (or the other way around: the button should be invisible if all sub-project have linked project).
Last but not least: the check on sub BR should be multi-level.

Let me know if this is clear

**Comment by sudhir-serpentcs** [comment]
@elicoidal 
improved  sub-BR code please review it.

**Comment by elicoidal** [comment]
@sudhir-serpentcs still not working as expected.
In my example:
* parent: SO stage
* Child1 (in SO stage)
* Child 2 (in draft stage)

Expected output is that 
* Parent Project is created (OK)
* Child 1 project is created (not OK)
* Child 2 project is not created (OK)
* Button is still visible in Parent (not OK) and I should still be able to relaunch the project generation fromthe parent for all child BR

**Comment by elicoidal** [comment]
@sudhir-serpentcs Can I retest this?

**Comment by YogeshMahera-SerpentCS** [comment]
@elicoidal 
Expected output is that

- Parent Project is created (OK)
- Child 1 project is created (not OK) **----> I guess you missed it will be OK because Child 1 is on SO stage**.
- Child 2 project is not created (OK)
- Button is still visible in Parent (not OK) and I should still be able to relaunch the project generation from the parent for all child BR.

**Comment by elicoidal** [comment]
@YogeshMahera-SerpentCS you are right!

**Comment by elicoidal** [comment]
Can I retest?

**Comment by YogeshMahera-SerpentCS** [comment]
> Button is still visible in Parent (not OK) and I should still be able to relaunch the project generation from the parent for all child BR.

@elicoidal I have one idea that we can create a new boolean (compute) field which will get true if all the child BRs have "Linked Project" (projects are created for all sub-brs) and then use this boolean field in attrs to hide the button in xml.
Let me know if you have better idea, I will go for that.

**Comment by martintorresvm** [comment]
> @elicoidal I have one idea that we can create a new boolean (compute) field which will get true if all the child BRs have "Linked Project" (projects are created for all sub-brs) and then use this boolean field in attrs to hide the button in xml.
Let me know if you have better idea, I will go for that.

@YogeshMahera-SerpentCS IMO is a good solution. But need to be stored, right?
How to call that field? 
I would go for `generate_project`

@elicoidal ?

**Comment by elicoidal** [comment]
Idea is OK.
Field name all_project_generated
:+1:

**Comment by YogeshMahera-SerpentCS** [comment]
@victormartinelicocorp 
calling method
`@api.depends('business_requirement_ids','business_requirement_ids.linked_project')`

**Comment by elicoidal** [comment]
Parent: SOA 
Child 1: SOA
Child 2: Draft
Just rechecked now and Child 1 BR in the current example has no generated Proejct although the status is SOA. 
Button Generate Project is visible but clicking on it doesnot do anything
Expected behavior: when clicking the button the first time should generate 2 projects
It should be later possible to change state of Child 2 to SOA and click on the Parent (or child 2) button to generate that project.
@YogeshMahera-SerpentCS

**Comment by YogeshMahera-SerpentCS** [comment]
@elicoidal 
Yes Same steps I have generated, I have taken some screenshot could please check it and please correct me if I am wrong.

**Comment by elicoidal** [comment]
@YogeshMahera-SerpentCS this is crystal now and I confirm the behavior!
Thanks for your patience

**Review by elicoidal — APPROVED** [review]

**Comment by elicoidal** [comment]
@victormartinelicocorp

File: business_requirement_deliverable_project/models/business.py [hunk 1]
```diff
@@ -44,6 +44,24 @@ def generate_projects_wizard(self):
         compute='_compute_planned_hour'
     )
 
+    all_project_generated = fields.Boolean(compute='compute_project_generate',
+                                           string='All Project Generated')
+
+    @api.depends('business_requirement_ids',
+                 'business_requirement_ids.linked_project')
+    def compute_project_generate(self):
```
Comment (martintorresvm) [inline business_requirement_deliverable_project/models/business.py]: Field name is `all_project_generated`
then this function should be named `compute_all_project_generated`

**Review by martintorresvm — CHANGES_REQUESTED** [review]

**Comment by YogeshMahera-SerpentCS** [comment]
@victormartinelicocorp Could  please review

**Comment by elicoidal** [comment]
@dreispt @pedrobaeza Can you help?

File: business_requirement_deliverable_project/wizard/generate_projects.py [hunk 2]
```diff
@@ -137,8 +137,9 @@ def generate_br_projects(self, parent_project, br, project_ids, task_ids):
         if self.for_childs:
             br_project = br_project or parent_project
             for child_br in br.business_requirement_ids:
-                self.generate_br_projects(
-                    br_project, child_br, project_ids, task_ids)
+                if child_br.state == 'stakeholder_approval':
```
Comment (pedrobaeza) [inline business_requirement_deliverable_project/wizard/generate_projects.py:138]: Do you have the opportunity later to generate the remaining projects?

File: business_requirement_deliverable_project/models/business.py [hunk 3]
```diff
@@ -44,6 +44,26 @@ def generate_projects_wizard(self):
...
+        compute='compute_all_project_generated',
+        string='All Project Generated'
+    )
+
+    @api.depends('business_requirement_ids',
+                 'business_requirement_ids.linked_project')
+    def compute_all_project_generated(self):
+        for rec in self:
+            if rec.business_requirement_ids:
+                if all(line.linked_project for line in
```
Comment (pedrobaeza) [inline business_requirement_deliverable_project/models/business.py]: `if all(rec.mapped('business_requirement_ids.linked_project'))`

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: business_requirement_deliverable_project/wizard/generate_projects.py (hunk 2 above)
Comment (elicoidal) [reply business_requirement_deliverable_project/wizard/generate_projects.py:138]: @pedrobaeza Yes: actually that is the point to allow the generation of the projects  in multiple steps (not all BR are ready at the same time).
Requirement is to be able to launch the wizard any time as the button is not made invisible until all projects are generated

**Comment by elicoidal** [comment]
@sudhir-serpentcs

**Comment by YogeshMahera-SerpentCS** [comment]
Improved Changes

**Review by elicoidal — APPROVED** [review]

**Comment by elicoidal** [comment]
@victormartinelicocorp @pedrobaeza

**Review by martintorresvm — APPROVED** [review]

**Comment by elicoidal** [comment]
@pedrobaeza can we move forward?

**Comment by pedrobaeza** [comment]
Please solve conflicts

**Comment by elicoidal** [comment]
@YogeshMahera-SerpentCS Can you have a look?

**Comment by YogeshMahera-SerpentCS** [comment]
Resolved conflicts

**Comment by elicoidal** [comment]
thanks: moving forward

**Pull request merged**